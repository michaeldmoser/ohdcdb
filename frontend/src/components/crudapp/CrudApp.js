import React, { useState } from 'react';
import { Route, Routes } from 'react-router-dom';

import Context from './context';

import {
    ListOfRecords,
    RecordDetail,
    RecordWithSearchResults,
} from './containers';

const CrudApp = ({ children, useGetListQuery, useGetRecordQuery, ...rest }) => {
    const [search, setSearch] = useState('');
    const [recordId, setRecordId] = useState(null);
    const listQuery = useGetListQuery(search);
    const detailQuery = useGetRecordQuery(recordId, { skip: !recordId });

    const kids = React.Children.toArray(children);
    const recordlist = kids.find(({ type }) => type === ListOfRecords);
    const recordview = kids.find(({ type }) => type === RecordDetail);

    return (
        <Context.Provider
            value={{
                search,
                setSearch,
                recordId,
                setRecordId,
                listQuery,
                detailQuery,
                ...rest,
            }}
        >
            <section className='container-fluid data-view h-100 overflow-hidden'>
                {/* <PropertiesHeader search={search} setSearchQuery={setSearch} /> */}
                <div className='card'>
                    <div className='row h-100 overflow-hidden'>
                        <Routes>
                            {/* <Route
                                index
                                element={<ListOnly>{recordlist}</ListOnly>}
                            /> */}
                            <Route
                                path='*'
                                element={
                                    <RecordWithSearchResults
                                        listview={recordlist}
                                    >
                                        {recordview}
                                    </RecordWithSearchResults>
                                }
                            >
                                {/* <Route path='add' element={<AddProperty />} /> */}
                                <Route path=':recordId' element={recordview} />
                                {/* <Route
                                    path=':recordId/edit'
                                    element={<EditProperty />}
                                /> */}
                            </Route>
                        </Routes>
                    </div>
                </div>
            </section>
        </Context.Provider>
    );
};

export default CrudApp;
