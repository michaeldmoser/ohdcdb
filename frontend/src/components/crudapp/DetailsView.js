import React, { useContext } from 'react';
import { Nav } from 'react-bootstrap';
import { NavLink } from 'react-router-dom';

import Context from './context';

export const ActionsMenu = () => {
    return (
        <nav className='ms-auto'>
            <Nav as='ul'>
                <Nav.Item as='li'>
                    <NavLink
                        to='edit'
                        className='btn btn-primary'
                        role='button'
                    >
                        Edit
                    </NavLink>
                </Nav.Item>
            </Nav>
        </nav>
    );
};

export const Header = ({ children, ...props }) => {
    return <h4>{children}</h4>;
};

export const Body = ({ children, ...props }) => {
    return (
        <div className='card-body'>
            <dl className='row'>{children}</dl>
        </div>
    );
};

export const Field = ({ label, value, ...props }) => {
    return (
        <>
            <dt className='col-sm-3'>{label}</dt>
            <dd className='col-sm-9'>{value}</dd>
        </>
    );
};

export const DetailsView = ({ children }) => {
    const {
        detailQuery: { data, isLoading },
    } = useContext(Context);
    if (isLoading) return <h4>Loading...</h4>;

    if (!data) return <div>Not found</div>;

    const kids = React.Children.toArray(children);
    const header = kids.find(({ type }) => type === Header);
    const body = kids.find(({ type }) => type === Body);

    return (
        <article className='detail-view'>
            <header className='d-flex'>
                {header}
                <ActionsMenu />
            </header>
            {body}
        </article>
    );
};

export default DetailsView;
