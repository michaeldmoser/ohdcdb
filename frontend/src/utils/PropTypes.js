import PropTypes from 'prop-types';

function isComponent(componentType) {
    const checkComponent = (
        propValue,
        key,
        componentName,
        location,
        propFullName
    ) => {
        if (!propValue[key].type === componentType) {
            return new Error(
                'Invalid prop `' +
                    propFullName +
                    '` supplied to' +
                    ' `' +
                    componentName +
                    '`. Validation failed. Should be a '
            );
        }
    };

    return checkComponent;
}

export default Object.assign({}, PropTypes, { isComponent });
