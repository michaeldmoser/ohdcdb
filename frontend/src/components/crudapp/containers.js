import { Children } from 'react';
import { useContext, useEffect } from 'react';
import { useParams, Outlet } from 'react-router-dom';
import PropTypes from 'prop-types';

import Context from './context';

export const RecordDetail = ({ children }) => {
    const { recordId } = useParams();
    const { detailQuery, setRecordId } = useContext(Context);

    useEffect(() => {
        setRecordId(recordId);
    }, [recordId, setRecordId]);

    return children(detailQuery);
};

RecordDetail.propTypes = {
    children: PropTypes.func.isRequired,
};

export const ListOfRecords = ({ children }) => {
    const context = useContext(Context);
    const { recordsList, recordsAreLoading } = context;

    return children({ data: recordsList, isLoading: recordsAreLoading });
};

export const RecordWithSearchResults = ({ listview, children }) => {
    return (
        <>
            <div className='col-3 d-none d-lg-block h-100 overflow-scroll'>
                {listview}
            </div>
            <div className='col-12 col-lg-9'>
                <Outlet />
            </div>
        </>
    );
};
