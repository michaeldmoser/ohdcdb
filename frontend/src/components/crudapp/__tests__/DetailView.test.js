import React from 'react';
import { render } from 'testing/library';

import { queryResult, database } from './utils';

import CrudApp from '../index';
import { ListOfRecords, RecordDetail } from '../containers';
import DetailsView, { Body, Field, Header } from '../DetailsView';

describe('Test displaying a record', () => {
    const useGetListQuery = (search) => {
        return queryResult(database);
    };

    const useGetRecordQuery = (recordId) => {
        return queryResult([1]);
    };

    it('should display the details of a record', async () => {
        const details = render(
            <CrudApp {...{ useGetListQuery, useGetRecordQuery }}>
                <ListOfRecords>
                    {(props) => {
                        return <div />;
                    }}
                </ListOfRecords>
                <RecordDetail>
                    {({ data, isLoading }) => {
                        return (
                            <DetailsView>
                                <Body>
                                    <Field value={data.title} label='Title' />
                                    <Field
                                        value={data.description}
                                        label='Description'
                                    />
                                </Body>
                                <Header>{data.title}</Header>
                            </DetailsView>
                        );
                    }}
                </RecordDetail>
            </CrudApp>,
            {
                initialEntries: ['/2'],
            }
        );

        expect(details).toMatchSnapshot();
    });
});
