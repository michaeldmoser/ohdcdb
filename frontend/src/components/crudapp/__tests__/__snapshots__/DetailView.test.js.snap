// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test displaying a record should display the details of a record 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="container-fluid data-view h-100 overflow-hidden"
      >
        <div
          class="card"
        >
          <div
            class="row h-100 overflow-hidden"
          >
            <div
              class="col-3 d-none d-lg-block h-100 overflow-scroll"
            >
              <div />
            </div>
            <div
              class="col-12 col-lg-9"
            >
              <article
                class="detail-view"
              >
                <header
                  class="d-flex"
                >
                  <h4 />
                  <nav
                    class="ms-auto"
                  >
                    <ul
                      class="nav"
                    >
                      <li
                        class="nav-item"
                      >
                        <a
                          class="btn btn-primary"
                          href="/2/edit"
                          role="button"
                        >
                          Edit
                        </a>
                      </li>
                    </ul>
                  </nav>
                </header>
                <div
                  class="card-body"
                >
                  <dl
                    class="row"
                  >
                    <dt
                      class="col-sm-3"
                    >
                      Title
                    </dt>
                    <dd
                      class="col-sm-9"
                    />
                    <dt
                      class="col-sm-3"
                    >
                      Description
                    </dt>
                    <dd
                      class="col-sm-9"
                    />
                  </dl>
                </div>
              </article>
            </div>
          </div>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="container-fluid data-view h-100 overflow-hidden"
    >
      <div
        class="card"
      >
        <div
          class="row h-100 overflow-hidden"
        >
          <div
            class="col-3 d-none d-lg-block h-100 overflow-scroll"
          >
            <div />
          </div>
          <div
            class="col-12 col-lg-9"
          >
            <article
              class="detail-view"
            >
              <header
                class="d-flex"
              >
                <h4 />
                <nav
                  class="ms-auto"
                >
                  <ul
                    class="nav"
                  >
                    <li
                      class="nav-item"
                    >
                      <a
                        class="btn btn-primary"
                        href="/2/edit"
                        role="button"
                      >
                        Edit
                      </a>
                    </li>
                  </ul>
                </nav>
              </header>
              <div
                class="card-body"
              >
                <dl
                  class="row"
                >
                  <dt
                    class="col-sm-3"
                  >
                    Title
                  </dt>
                  <dd
                    class="col-sm-9"
                  />
                  <dt
                    class="col-sm-3"
                  >
                    Description
                  </dt>
                  <dd
                    class="col-sm-9"
                  />
                </dl>
              </div>
            </article>
          </div>
        </div>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
